docs/source/distributed.pipelining.rst

```diff
@@ -4,182 +4,310 @@
 
-We used ``annotate_split_points`` to specify that the code should be split and the end of ``layer0`` and ``layer1``. Our code has thus been split into *three* pipeline stages. Our library also provides ``SplitPoint.BEGINNING`` if a user wants to split before certain annotation point.
+The "model partitions" are represented by submodules (``submod_0``,
+``submod_1``), each of which are reconstructed with original model operations
+and hierarchies.  In addition, a "root-level" ``forward`` function is
+reconstructed to capture the data flow between those partitions. Such data flow
+will be replayed by the pipeline runtime later, in a distributed fashion.
 
-While the ``annotate_split_points`` API gives users a way to specify the split points without modifying the model, our library also provides an API for in-model annotation: ``pipe_split()``. For details, you can read `this example <https://github.com/pytorch/PiPPy/blob/main/test/test_pipe.py>`_.
+The ``Pipe`` object provides a method for retrieving the "model partitions":
 
-This covers the basic usage of the ``Pipe`` API. For more information, please see the documentation.
-
-Using ``PipelineSchedule`` for Execution
-========================================
+.. code-block:: python
 
-After transforming the model into a ``Pipe`` representation, we can run its stages in a distributed *runtime*. This can be done in two steps:
-* instantiate a ``PipelineStage`` from a stage module of ``Pipe``;
-* run the ``PipelineStage`` according to a ``PipelineSchedule``.
+  stage_mod : nn.Module = pipe.get_stage_module(stage_idx)
 
-First off, let us instantiate a ``PipelineStage`` instance:
+You can also create a distributed stage runtime on a device using ``Pipe``:
 
 .. code-block:: python
 
-  # We are using `torchrun` to run this example with multiple processes.
-  # `torchrun` defines two environment variables: `RANK` and `WORLD_SIZE`.
-  rank = int(os.environ["RANK"])
-  world_size = int(os.environ["WORLD_SIZE"])
+  from torch.distributed.pipelining import PipelineStage
 
-  # Initialize distributed environment
-  import torch.distributed as dist
-  dist.init_process_group(rank=rank, world_size=world_size)
+  stage = PipelineStage(pipe, stage_idx, device)
 
-  # Pipeline stage is our main pipeline runtime. It takes in the pipe object,
-  # the rank of this process, and the device.
-  from torch.distributed.pipelining import PipelineStage
-  stage = PipelineStage(pipe, rank, device)
+.. note::
+  The ``pipeline`` frontend uses a tracer (``torch.export``) to capture your
+  model into a single graph. If your model is not single-graphable, you can use
+  our manual frontend below.
+
+Frontend 2: ``ManualPipelineStage`` -- If You Already Have the Module for Each Stage
+====================================================================================
+
+If you already have the module for each stage, you can skip the pipeline split
```

Comment:
I think I would also like to give suggestions for how to write a 'pipeline-friendly' model to use the manual stage.

The basic guideline, which derives from your point above about 'partitioning the execution', is to write your top-level model forward() in such a way that it will run correctly if child modules are 'deleted' or 'set to none'.  And point to the titan llama model as an example.